### PR TITLE
Vet clinic database: add "join table" for visits

### PR DESCRIPTION
**Pull request by shm04**
### In this pull request I add the following changes:

- Create a table named vets with the following columns:
  - id: integer (set it as autoincremented PRIMARY KEY)
  - name: string
  - age: integer
  - date_of_graduation: date
- There is a many-to-many relationship between the tables species and vets: a vet can specialize in multiple species, and a species can have multiple vets specialized in it. Create a "join table" called specializations to handle this relationship.
- There is a many-to-many relationship between the tables animals and vets: an animal can visit multiple vets and one vet can be visited by multiple animals. Create a "join table" called visits to handle this relationship, it should also keep track of the date of the visit.
- Insert the following data for vets:
  - Vet William Tatcher is 45 years old and graduated Apr 23rd, 2000.
  - Vet Maisy Smith is 26 years old and graduated Jan 17th, 2019.
  - Vet Stephanie Mendez is 64 years old and graduated May 4th, 1981.
  - Vet Jack Harkness is 38 years old and graduated Jun 8th, 2008.
- Insert the following data for specializations:
  - Vet William Tatcher is specialized in Pokemon.
  - Vet Stephanie Mendez is specialized in Digimon and Pokemon.
  - Vet Jack Harkness is specialized in Digimon.
- Insert the following data for visits:
  - Agumon visited William Tatcher on May 24th, 2020.
  - Agumon visited Stephanie Mendez on Jul 22th, 2020.
  - Gabumon visited Jack Harkness on Feb 2nd, 2021.
  - Pikachu visited Maisy Smith on Jan 5th, 2020.
  - Pikachu visited Maisy Smith on Mar 8th, 2020.
  - Pikachu visited Maisy Smith on May 14th, 2020.
  - Devimon visited Stephanie Mendez on May 4th, 2021.
  - Charmander visited Jack Harkness on Feb 24th, 2021.
  - Plantmon visited Maisy Smith on Dec 21st, 2019.
  - Plantmon visited William Tatcher on Aug 10th, 2020.
  - Plantmon visited Maisy Smith on Apr 7th, 2021.
  - Squirtle visited Stephanie Mendez on Sep 29th, 2019.
  - Angemon visited Jack Harkness on Oct 3rd, 2020.
  - Angemon visited Jack Harkness on Nov 4th, 2020.
  - Boarmon visited Maisy Smith on Jan 24th, 2019.
  - Boarmon visited Maisy Smith on May 15th, 2019.
  - Boarmon visited Maisy Smith on Feb 27th, 2020.
  - Boarmon visited Maisy Smith on Aug 3rd, 2020.
  - Blossom visited Stephanie Mendez on May 24th, 2020.
  - Blossom visited William Tatcher on Jan 11th, 2021.
- Write queries to answer the following:
  - Who was the last animal seen by William Tatcher?
image
  - How many different animals did Stephanie Mendez see?
image
  - List all vets and their specialties, including vets with no specialties.
image
  - List all animals that visited Stephanie Mendez between April 1st and August 30th, 2020.
image
  - What animal has the most visits to vets?
image
  - Who was Maisy Smith's first visit?
image
  - Details for most recent visit: animal information, vet information, and date of visit.
image
  - How many visits were with a vet that did not specialize in that animal's species?
image
  - What specialty should Maisy Smith consider getting? Look for the species she gets the most.
image

### CODE REVIEW

File: queries.sql [hunk 1]
```diff
@@ -51,4 +51,29 @@ SELECT owners.full_name, animals.name FROM owners LEFT JOIN animals ON owners.id
...
+       vets.name AS vet_name,
+       vets.age AS vet_age,
+       vets.date_of_graduation AS vet_date_of_graduation,
+       visits.visit_date
+FROM visits
+JOIN animals ON visits.animal_id = animals.id
+JOIN vets ON visits.vet_id = vets.id
+JOIN species ON animals.species_id = species.id
+ORDER BY visits.visit_date DESC LIMIT 1;
+SELECT COUNT(*) AS num_visits_not_specialized FROM visits JOIN animals ON visits.animal_id = animals.id JOIN vets ON visits.vet_id = vets.id LEFT JOIN specializations ON vets.id = specializations.vet_id AND animals.species_id = specializations.species_id WHERE specializations.vet_id IS NULL;
```
Comment (Munsa1) [inline queries.sql:77]: - [ ] Here we can see that `Maisy Smith` has no specialty hence every animal that visited her should be counted, and those animals are 9 in total, this excluding the other animals that have a specialty but saw animals they were not specialized in, So we can see that the query output is incorrect. kindly fix this.

Given data
image

Your output
image

File: queries.sql [hunk 2]
```diff
@@ -51,4 +51,29 @@ SELECT owners.full_name, animals.name FROM owners LEFT JOIN animals ON owners.id
...
 SELECT animals.name FROM animals JOIN owners ON animals.owner_id = owners.id WHERE owners.full_name = 'Dean Winchester' AND animals.escape_attempts = 0;
-SELECT owners.full_name, COUNT(animals.id) AS num_owned FROM owners LEFT JOIN animals ON owners.id = animals.owner_id GROUP BY owners.id, owners.full_name ORDER BY num_owned DESC;
\ No newline at end of file
+SELECT owners.full_name, COUNT(animals.id) AS num_owned FROM owners LEFT JOIN animals ON owners.id = animals.owner_id GROUP BY owners.id, owners.full_name ORDER BY num_owned DESC;
+
+SELECT animals.name FROM animals JOIN visits ON animals.id = visits.animal_id JOIN vets ON visits.vet_id = vets.id WHERE vets.name = 'William Tatcher' ORDER BY visits.visit_date DESC;
+SELECT COUNT(DISTINCT visits.animal_id) AS num_animals_seen FROM visits JOIN vets ON visits.vet_id = vets.id WHERE vets.name = 'Stephanie Mendez';
+SELECT vets.name, species.name FROM vets LEFT JOIN specializations ON vets.id = specializations.vet_id LEFT JOIN species ON specializations.species_id = species.id ORDER BY vets.id, species.id;
```
Comment (Munsa1) [inline queries.sql:58]: - [ ] Kindly list the correct values for the vets and their specialties.

List all vets and their specialties, including vets with no specialties.
Answer:

William Tatcher (Specialty: Pokemon)
Maisy Smith (No Specialties)
Stephanie Mendez (Specialties: Pokemon, Digimon)
Jack Harkness (Specialty: Digimon)



Your output
image

**Review by Munsa1 — CHANGES_REQUESTED** [review]
Hi @shm04, 

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!
<FEEL FREE TO CHANGE THIS MESSAGE TO YOUR PERSONAL WORDS>

## Highlights
Provided Screenshots For Every Query :heavy_check_mark: 
Correct Git Flow :heavy_check_mark: 
Descriptive ReadMe File :heavy_check_mark: 

## Required Changes ♻️
_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

You can also consider:
n/a

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: queries.sql [hunk 3]
```diff
@@ -51,4 +51,29 @@ SELECT owners.full_name, animals.name FROM owners LEFT JOIN animals ON owners.id
...
+SELECT animals.name FROM animals JOIN visits ON visits.animal_id = animals.id WHERE visits.visit_date BETWEEN '2020-04-01' AND '2020-08-30';
```
[the remark this answers is not in the document]
Comment (shm04) [reply queries.sql:59]: In this, I think the correct answer is Agumon and Blossom, Pikachu didn't go with Stephanie and Plantom neither

File: queries.sql [hunk 4]
```diff
@@ -51,4 +51,29 @@ SELECT owners.full_name, animals.name FROM owners LEFT JOIN animals ON owners.id
...
+       species.name AS species_name,
+       vets.name AS vet_name,
+       vets.age AS vet_age,
+       vets.date_of_graduation AS vet_date_of_graduation,
+       visits.visit_date
+FROM visits
+JOIN animals ON visits.animal_id = animals.id
+JOIN vets ON visits.vet_id = vets.id
+JOIN species ON animals.species_id = species.id
+ORDER BY visits.visit_date DESC LIMIT 1;
```
[the remark this answers is not in the document]
Comment (shm04) [reply queries.sql:76]: This also correct because Blossom went on January and Devimon on May

**Review by Munsa1 — APPROVED** [review]
Hi @shm04, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 

## Highlights
Made Requested Changes :heavy_check_mark: 


## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**